pkg/kubecost/allocation.go

```diff
@@ -2260,8 +2260,46 @@ func (asr *AllocationSetRange) Accumulate() (*AllocationSet, error) {
 	return allocSet, nil
 }
 
-// TODO accumulate into lower-resolution chunks of the given resolution
-// func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) *AllocationSetRange
+// AccumulateBy sums AllocationSets based on the resolution given. The resolution given is subject to the scale used for the AllocationSets.
+// If the requested resolution is smaller than the window of an AllocationSet then the resolution will default to the duration of a set.
+// Resolutions larger than the duration of the entire AllocationSetRange will default to the duration of the range.
+func (asr *AllocationSetRange) AccumulateBy(resolution time.Duration) (*AllocationSetRange, error) {
+	allocSetRange := &AllocationSetRange{allocations: []*AllocationSet{}}
+	var allocSet *AllocationSet
+	var currAccumulatedSum time.Duration
+	var err error
+
+	asr.Lock()
+	defer asr.Unlock()
+
+	// // use window() with no lock
+	if asr.window().IsEmpty() {
+		return asr, nil
+	}
+
+	asrWindow := asr.window()
+
+	// check as.window and accumulate sets until sum of window durations == time.duration wanted -> add accumulated set to allocSetRange
+	for _, as := range asr.allocations {
+		allocSet, err = allocSet.accumulate(as)
+		if err != nil {
+			return nil, err
+		}
+
+		if allocSet != nil {
+			currAccumulatedSum += allocSet.Window.Duration()
+
+			// check if end of asr to sum the final set too
+			if currAccumulatedSum >= resolution || NewWindow(nil, asrWindow.end).Equal(NewWindow(nil, allocSet.Window.end)) {
```

Comment:
I think you could replace the `NewWIndow().Equal(NewWindow())` check with just `asrWindow.end == allocSet.Window.end`